fifo/fifo_group.go

```diff
@@ -266,3 +284,83 @@ func groupFromJSON(b []byte) (*parse.Result, error) {
 
 	return parse.NewResult(g, msg.Scope)
 }
+
+// flattenedResponseMods if a Group has a request modifier that is another Group
+// and the aggregateErrors settings match between the both groups the other group's modifiers are inlined.
+// This is applied recursively so that it produces the most flat Group of all modifiers.
+func (g *Group) flattenedResponseMods() ([]martian.ResponseModifier, error) {
+	g.resmu.Lock()
+	defer g.resmu.Unlock()
+
+	var resmods []martian.ResponseModifier
+	for _, m := range g.resmods {
+		if mm, ok := m.(*Group); ok {
+			if g.aggregateErrors != mm.aggregateErrors {
+				return nil, fmt.Errorf("aggregateErrors settings differ: %v, %v", g.aggregateErrors, mm.aggregateErrors)
+			}
+			flattened, err := mm.flattenedResponseMods()
+			if err != nil {
+				return nil, err
+			}
+			resmods = append(resmods, flattened...)
+			continue
+		}
+		resmods = append(resmods, m)
+	}
+
+	return resmods, nil
+}
+
+// flattenedRequestMods if a Group has a request modifier that is another Group
+// and the aggregateErrors settings match between the both groups the other group's modifiers are inlined.
+// This is applied recursively so that it produces the most flat Group of all modifiers.
+func (g *Group) flattenedRequestMods() ([]martian.RequestModifier, error) {
+	g.reqmu.Lock()
+	defer g.reqmu.Unlock()
+
+	var reqmods []martian.RequestModifier
+	for _, m := range g.reqmods {
+		if mm, ok := m.(*Group); ok {
+			if g.aggregateErrors != mm.aggregateErrors {
+				return nil, fmt.Errorf("aggregateErrors settings differ: %v, %v", g.aggregateErrors, mm.aggregateErrors)
+			}
+			flattened, err := mm.flattenedRequestMods()
+			if err != nil {
+				return nil, err
+			}
+			reqmods = append(reqmods, flattened...)
+			continue
+		}
+		reqmods = append(reqmods, m)
+	}
+
+	return reqmods, nil
+}
+
+// ToImmutable creates ImmutableGroup from existing Group, additionally flattening all the modifiers.
+func (g *Group) ToImmutable() (*ImmutableGroup, error) {
```

Comment:
The relationship between ToImmutable and flattening and is wrong and it affects the API now.
The primary concern is making Group immutable recursively.
This can always be done no need to return the error.
When we have the immutable group, in some cases, we may inline it.

This can be done in a single for loop, without errors logging etc.